tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/cluster/distribution/AMQPMessageRedistributionTest.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.activemq.artemis.core.settings.impl.AddressSettings;
+import org.apache.activemq.artemis.protocol.amqp.broker.ProtonProtocolManagerFactory;
+import org.apache.qpid.jms.JmsConnectionFactory;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.jms.Connection;
+import javax.jms.Message;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+public class AMQPMessageRedistributionTest extends ClusterTestBase {
+
+   final String queue = "exampleQueue";
+   final String broker0 = "amqp://localhost:61616";
+   final String broker1 = "amqp://localhost:61617";
+
+   @Override
+   @Before
+   public void setUp() throws Exception {
+      super.setUp();
+      start();
+   }
+
+   private void start() throws Exception {
+      setupServers();
+
+      setRedistributionDelay(0);
+   }
+
+   protected boolean isNetty() {
+      return true;
+   }
+
+   @Test
+   public void testMessageRedistributionWithoutDupAMQP() throws Exception {
+      setupCluster(MessageLoadBalancingType.ON_DEMAND);
+
+      startServers(0, 1);
+
+      setupSessionFactory(0, isNetty());
+      setupSessionFactory(1, isNetty());
+
+      createQueue(0, queue, queue, null, true, null, null, RoutingType.ANYCAST);
+      createQueue(1, queue, queue, null, true, null, null, RoutingType.ANYCAST);
+
+      waitForBindings(0, queue, 1, 0, true);
+      waitForBindings(1, queue, 1, 0, true);
+
+      waitForBindings(0, queue, 1, 0, false);
+      waitForBindings(1, queue, 1, 0, false);
+
+      final int NUMBER_OF_MESSAGES = 20;
+
+      JmsConnectionFactory factory = new JmsConnectionFactory(broker0);
+      Connection connection = factory.createConnection();
+      Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+      MessageProducer producer = session.createProducer(session.createQueue(queue));
+
+      for (int i = 0; i < NUMBER_OF_MESSAGES; i++) {
+         producer.send(session.createTextMessage("hello " + i));
+      }
+      connection.close();
+
+      receiveOnBothNodes(NUMBER_OF_MESSAGES);
+   }
+
+   private void receiveOnBothNodes(int NUMBER_OF_MESSAGES) throws Exception {
+      receiveBroker(broker1, 1);
+      receiveBroker(broker0, 1);
+      receiveBrokerAll(broker1, NUMBER_OF_MESSAGES - 2);
+   }
+
+   private void receiveBrokerAll(String brokerurl, int num) throws Exception {
+      JmsConnectionFactory factory = new JmsConnectionFactory(brokerurl);
+      Connection connection = factory.createConnection();
+      Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+      MessageConsumer consumer = session.createConsumer(session.createQueue(queue));
+      connection.start();
+
+      for (int i = 0; i < num; i++) {
+         TextMessage msg = (TextMessage) consumer.receive(5000);
+         assertNotNull(msg);
+      }
+      Message msg = consumer.receive(2000);
```

Comment:
Please... do not receive(2 seconds) and assert for null

if you expect it to be null, please receiveNoWait

Reply:
but I already changed it here, no need to ammend.